File: epi-cpp/14/RenderCalender.cpp
#include <vector>
#include <algorithm>

namespace RenderCalender {
    using namespace std;
    class Interval {
    public:
        int start;
        int end;
        Interval (int start, int end){
            this-> start = start;
            this->end = end;
        }
    };


    class EndPoint {
    public:
        int time;
        bool isStart;
        EndPoint (int time, int isStart){
            this->time = time;
            this->isStart = isStart;
        }
    };

    int solve (vector<Interval> intervalList) {
        vector<EndPoint> epList;
        for (auto interval: intervalList){
            epList.push_back(EndPoint(interval.start, true));
            epList.push_back(EndPoint(interval.end, false));
        }

        sort (epList.begin(), epList.end(), [](EndPoint a, EndPoint b){
            return (a.time < b.time);
        });

        int counter = 0;
        int ans = 0;
        for (auto ep: epList) {
            if (ep.isStart)
                counter++;
            else
                counter--;
            ans = max(ans, counter);
        }

        return ans;
    }

    void test () {
        vector<Interval> input = {
                Interval(1,5), Interval(6,10), Interval(11,13), Interval(14,15), Interval(2,7),
                Interval(8,9), Interval(12,15), Interval(4,5), Interval(9,17)
        };
        cout << solve (input);
        return;
    }
};

File: epi-java/src/sorting/RenderCalender.java
package sorting;


import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import util.EndPoint;
import util.Interval;


public class RenderCalender {

    public int solve(List<Interval> intervals) {
        List<EndPoint> endPoints = new ArrayList<>();
        for (Interval i : intervals) {
            endPoints.add(new EndPoint(i.begin, true));
            endPoints.add(new EndPoint(i.end, false));
        }

        Collections.sort(endPoints);
        int ans = Integer.MIN_VALUE;
        int counter = 0;
        for (EndPoint e : endPoints) {
            if (e.isStart) {
                counter += 1;
                ans = Math.max(ans, counter);
            } else {
                counter -= 1;
            }
        }
        return ans;
    }

    public void test() {
        Interval[] t = {Interval.make(1, 5), Interval.make(2, 7), Interval.make(4, 5),
                Interval.make(6, 10), Interval.make(8, 9), Interval.make(9, 17), Interval.make(11, 13),
                Interval.make(12, 15), Interval.make(14, 15)};
        System.out.println(solve(Arrays.asList(t)));
    }
}
